### PR TITLE
RE flowsheet

### PR DESCRIPTION
**Pull request by dguittet**
## Summary/Motivation:
Adds a flowsheet with components organized as shown here:

image

 - `air_feed` is currently left free so that the correct stoichiometric ratio of air to hydrogen can be found respecting the temperature bounds of the h2_turbine. The `initialize_model` fxn guesses at a good starting value.
 - flowsheet is set up to use the battery charging and h2_tank outlet flow rate as control variables

Tests show that flowsheet solves for a select range of design variables

## Changes proposed in this PR:
- Addition of RE_flowsheet and tests
- Contains hydrogen tank model updates that may replicate (or cause conflicts with) #61 
- Increases pressure upper bound in hturbine_ideal_vap.py 
- Changes to initialization in hydrogen_turbine_unit.py
- Battery model variables now indexed by time and add `initialize`
- `initialize` and `report` for PEM

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.md and COPYRIGHT.md file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Review by klfrick2 — APPROVED** [review]

**Comment by radhakrishnatg** [comment]
@dguittet A quick question. The header is formatted in a lot files. Is the modified header consistent with the actual header? Also, I noticed that a couple of files in the nuclear case study have been modified slightly. I changed those files in #84. So, to avoid merge conflicts, shall we wait until the PR is merged? I will ask Ludovico to merge the PR as soon as possible.

**Comment by dguittet** [comment]
@radhakrishnatg I have been running `scripts/annotate_source` and so I'm not sure why there are a different number of hashes on some headers.

I have pulled in the changes from the nuclear case study and resolved most of the conflicts.

There is one remaining that is due to changes I've made to the elec_splitter unit model: https://github.com/gmlc-dispatches/dispatches/pull/60/files#diff-c1b5c3ceb35753859359150476ec7b4c77479ef1aec251cc5956a3b4f7525c0b

I have linearized this unit model by making the split_fraction an Expression rather than a Variable, so now the only constraint is electricity conservation. Since the only quantity to balance is electricity, we don't really need a Variable to assure that different quantities have the same split fraction.

This is causing a test break for the nuclear flowsheet where it tries to fix this split_fraction Expression: https://github.com/gmlc-dispatches/dispatches/runs/6098363743?check_suite_focus=true#step:7:84

**Comment by radhakrishnatg** [comment]
@dguittet Sometime back, I added the header by running `addheader -c .addheader.yml` in the command window. @lbianchi-lbl Is the header generated by `addheader -c .addheader.yml` and `scripts/annotate_source` different? If yes, which one is the recommended option?

Regarding the `elec_splitter` model, I agree that split fraction is not needed for electricity balance. It is good to have the entire model in linear form. But declaring split fraction as an expression would cause the issue you noticed. Users won't be able to `fix` the value. Instead, they need to add a constraint `...split_frac_constraint = Constraint(expr=...split_frac[...] == 0.2)`. If they add such a constraint, the model then becomes nonlinear, because pyomo will not rearrange the constraint in linear form. 

I'm not sure what might be appealing for the users, declaring split fraction as a variable and fixing it or adding a constraint. Some users would like to have split fractions as variables in the model. So, this is my suggestion. Let me know your thoughts. Add one more input option
```python
    CONFIG.declare(
        "add_split_fraction_vars",
        ConfigValue(
            domain=In([False, True]),
            default=False,
            description="Add split fraction variables. Set it to True if these variables are needed"
        )
    )
```
And in the `build` method, you can add the split fraction variables and the constraints defining split fractions if the flag is true. This way, if the user is interested in adding these variables, they can do so by setting the flag true. 
```python
@self.Constraint(time, doc="Electricity balance")
        def sum_split(b, t):
            return b.electricity[t] == sum(getattr(b, o + "_elec")[t] for o in b.outlet_list)

if self.config.add_split_fraction_vars:
    self.split_fraction = Var(self.outlet_list,
                                  time,
                                  bounds=(0, 1),
                                  initialize=1.0 / len(self.outlet_list),
                                  doc="Split fractions for outlet streams")

    @self.Constraint(time, self.outlet_list, doc="Split fraction definition")
        def split_fraction_definition(b, t, o):
            outlet_obj = getattr(b, o + "_elec")
            return outlet_obj[t] == b.split_fraction[o, t] * b.electricity[t]
```
Please keep the electricity balance equation the way you have it right now.

If you are okay with this, in the nuclear flowsheet file (https://github.com/gmlc-dispatches/dispatches/blob/e7f5e98f2e4c59ebbe963b22eb89bc9cf361b091/dispatches/models/nuclear_case/flowsheets/nuclear_flowsheet.py), you just need to make this change:
```python
    m.fs.np_power_split = ElectricalSplitter(default={
        "num_outlets": 2,
        "outlet_list": ["np_to_grid", "np_to_pem"],
        "add_split_fraction_vars": True})
```
This should fix the test without breaking anything else.

**Comment by dguittet** [comment]
@radhakrishnatg Thanks for that! I was thinking the same. I've implemented your code and the tests are passing.

File: dispatches/models/renewables_case/elec_splitter.py
Comment (radhakrishnatg) [inline dispatches/models/renewables_case/elec_splitter.py:232]: Here, we are changing the degrees of freedom by fixing/unfixing variables to make it a square problem. However, those variables are not reverted to their original state after the initialization. Either we need to print a warning saying that the state of the variables has been changed (so that the users are aware of it), or we need to get it back to their original state.

File: dispatches/models/renewables_case/wind_power.py
Comment (radhakrishnatg) [inline dispatches/models/renewables_case/wind_power.py:118]: I think, line 118 `self.setup_atb_turbine()` can be deleted. The method is called in line 126, so we can avoid the repetition.

File: dispatches/models/renewables_case/wind_power.py
Comment (radhakrishnatg) [inline dispatches/models/renewables_case/wind_power.py:169]: If the model has a single equation, then we don't have to use ipopt. We can just "set" the value of electricity from the equation:
```python
for t in self.flowsheet().config.time:
    self.electricity[t].value = self.system_capacity.value * self.capacity_factor[t]
```
This would save some computational time, especially if you need to call initialize method multiple times in the code.

File: dispatches/models/renewables_case/battery.py
Comment (radhakrishnatg) [inline dispatches/models/renewables_case/battery.py:181]: Could you explain why you are using `solve_indexed_blocks` instead of `opt.solve(self, tee=slc.tee)`? If I understand it correctly, the entire model is in a single block, so the latter should work.

File: dispatches/models/renewables_case/wind_battery_PEM_tank_turbine_LMP.py
Comment (radhakrishnatg) [inline dispatches/models/renewables_case/wind_battery_PEM_tank_turbine_LMP.py:296]: This works and there is no need to change it. I just wanted to note that this not a recommended practice. In general, variables from a parent block (in this case `m.pem_system_capacity`) are not recommended to be used in its child block (`blk_pem` in this case). This is because it can create issues especially if you want to solve the child block separately. To my knowledge, this is the recommended approach. Instead of adding the constraint to `blk_pem` object, add it to `m` i.e., the parent block which contains the variable `m.pem_system_capacity`.

File: dispatches/models/renewables_case/RE_flowsheet.py
Comment (radhakrishnatg) [inline dispatches/models/renewables_case/RE_flowsheet.py:190]: Here, I would recommend using `fix` (`m.fs.translator.outlet.mole_frac_comp[0, "hydrogen"]`) instead of adding a constraint. This way, the variable for the mole fraction of hydrogen does not appear in the optimization model that the solver sees. While they are mathematically the same, sometimes the solver works better with the former.

File: .gitignore
Comment (radhakrishnatg) [inline .gitignore:50]: I'm not sure what this does. Could you please double check with someone else if we can push this to the repo or not?

**Review by radhakrishnatg — COMMENTED** [review]
@dguittet I have reviewed the PR. Everything looks good. I have included a few minor comments. Let me know your thoughts on those. The Jupyter notebook currently just plots the results. From what I understand, the objective of the Jupyter notebook is demonstrate how we formulate the multiperiod optimization problem and solve it. My suggestion is to convert the `wind_battery_PEM_tank_turbine_LMP.py` script to Jupyter notebook. You do not have to formulate the problem for the whole year, since that would really long to solve (from what I remember from one of the earlier meetings). Just demonstrate the workflow with a week or one month's signal.

File: .gitignore
Comment (ksbeattie) [reply .gitignore:50]: I think this is ok.  We might even want to change the 2nd line here to `.ipynb_checkpoints/` so it picks up all of them.

**Comment by ksbeattie** [comment]
@lbianchi-lbl is going to look into how best to "unfix" all those minor header changes.

**Comment by lbianchi-lbl** [comment]
@dguittet @radhakrishnatg to revert the changes to the headers (so that e.g. other unrelated files don't show up as changes in this PR) you should be able to use the following commands:

```sh
addheader -c .addheader.yml --remove
addheader -c .addheader.yml
```

And then commit the changes.

File: dispatches/models/renewables_case/elec_splitter.py
Comment (dguittet) [reply dispatches/models/renewables_case/elec_splitter.py:231]: Doesn't the `to_json` in line 187 and the `from_json` in line 251 handle restoring the original state?

File: dispatches/models/renewables_case/wind_power.py
Comment (dguittet) [reply dispatches/models/renewables_case/wind_power.py]: done

File: dispatches/models/renewables_case/wind_battery_PEM_tank_turbine_LMP.py
Comment (dguittet) [reply dispatches/models/renewables_case/wind_battery_PEM_tank_turbine_LMP.py]: thanks I've moved all these constraints to the parent block:

```
    m.pem_max_p = Constraint(mp_wind_pem.pyomo_model.TIME,
                                rule=lambda b, t: blks[t].fs.pem.electricity[0] <= m.pem_system_capacity)
```

File: dispatches/models/renewables_case/battery.py
Comment (dguittet) [reply dispatches/models/renewables_case/battery.py]: fixed!

File: dispatches/models/renewables_case/RE_flowsheet.py
Comment (dguittet) [reply dispatches/models/renewables_case/RE_flowsheet.py]: fixed

**Comment by dguittet** [comment]
@radhakrishnatg @ksbeattie This PR is ready for re-review

File: dispatches/models/renewables_case/elec_splitter.py
Comment (radhakrishnatg) [reply dispatches/models/renewables_case/elec_splitter.py:231]: Sorry, Darice! I missed those two lines. I'm not familiar with all the options of `StoreSpec` class. In line 187, you are storing the list of variables that are fixed. Does the dictionary `istate` contain the variables that are not fixed? 
In line 251, you are passing this dictionary `istate` as input. I can see that it reverts the variables which are fixed by the user to their initial state. But how about the variables which were not fixed by the user but fixed by the initialization method? I'm assuming that the information about these variables will not be in the `istate` dictionary, so `to_json` function does not unfix these variables. Please correct me if my understanding is not correct.

**Review by radhakrishnatg — COMMENTED** [review]
@dguittet Everything looks good. I just have one question regarding the `to_json` function.

File: dispatches/models/renewables_case/elec_splitter.py
Comment (dguittet) [reply dispatches/models/renewables_case/elec_splitter.py:231]: No problem! I have checked those functions and even with `only_fixed=True`, the `from_json(self, sd=istate, wts=sp)` does restore previously unfixed variables back to unfixed state if the initialization fixes them. The DOF before and after initialization is the same in all the tests I have for elec_splitter.py.

Comment (radhakrishnatg) [reply dispatches/models/renewables_case/elec_splitter.py:231]: Alright. Thank you, Darice, for clarifying the clarification!

**Review by radhakrishnatg — APPROVED** [review]

File: dispatches/models/renewables_case/RE_flowsheet.py
Comment (xiangao1) [inline dispatches/models/renewables_case/RE_flowsheet.py:112]: Seems like the charging and discharging efficiency are fixed at 1 here.

Comment (dguittet) [reply dispatches/models/renewables_case/RE_flowsheet.py]: Changed to 0.98

Comment (xiangao1) [reply dispatches/models/renewables_case/RE_flowsheet.py]: Thanks, @dguittet ! Sorry to bother you again, but could you make this to 0.95 so it is consistent with this: https://github.com/xiangao1/dispatches/blob/a5d85fd8b5d42e7800f20cc5fb957965ae323b1a/dispatches/models/renewables_case/battery.py#L85.

Comment (dguittet) [reply dispatches/models/renewables_case/RE_flowsheet.py]: No problem, updated!

**Review by jghouse88 — APPROVED** [review]
Looks good!

**Pull request merged**
